File: app/src/main/java/me/pwcong/jpstart/network/baidu/service/impl/BaiduTranslateServiceImpl.java
package me.pwcong.jpstart.network.baidu.service.impl;

import com.blankj.utilcode.util.EncryptUtils;

import me.pwcong.jpstart.mvp.bean.BaiduTranslateBean;
import me.pwcong.jpstart.network.Api;
import me.pwcong.jpstart.network.baidu.BaiduTranslateApi;
import me.pwcong.jpstart.network.baidu.service.BaiduService;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by Pwcong on 2016/10/1.
 */

public class BaiduTranslateServiceImpl implements BaiduService.TranslateService {

    private static Retrofit instance = null;

    public static synchronized Retrofit getInstance() {

        if (instance == null) {

            instance = new Retrofit.Builder()
                    .baseUrl(Api.BAIDU_TRANSLATE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                    .build();

        }
        return instance;
    }

    @Override
    public void translate(String q, String from, String to, Subscriber<BaiduTranslateBean> subscriber) {

        int salt = (int) (Math.random() * 99999);
        String sign = EncryptUtils
                .encryptMD5ToString(Api.BAIDU_TRANSLATE_APPID + q + salt + Api.BAIDU_TRANSLATE_SECRETKEY).toLowerCase();

        getInstance().create(BaiduTranslateApi.class)
                .request(q, from, to, Api.BAIDU_TRANSLATE_APPID, salt, sign)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(subscriber);

    }
}

File: app/src/main/java/com/github/pwcong/jpstart/network/baidu/service/impl/BaiduTranslateServiceImpl.kt
package com.github.pwcong.jpstart.network.baidu.service.impl

import com.blankj.utilcode.util.EncryptUtils
import com.github.pwcong.jpstart.mvp.bean.BaiduTranslateBean
import com.github.pwcong.jpstart.network.baidu.api.BaiduTranslateApi
import com.github.pwcong.jpstart.network.baidu.service.BaiduService.TranslateService
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.util.Locale

class BaiduTranslateServiceImpl : TranslateService {
    override fun translate(
        q: String,
        from: String,
        to: String,
        subscriber: Subscriber<BaiduTranslateBean>
    ) {
        val salt = (Math.random() * 99999).toInt()
        val sign: String = EncryptUtils
            .encryptMD5ToString(BaiduTranslateApi.BAIDU_TRANSLATE_APPID + q + salt + BaiduTranslateApi.BAIDU_TRANSLATE_SECRETKEY)
            .lowercase(
                Locale.getDefault()
            )

        getInstance().create(
            BaiduTranslateApi::class.java
        )
            .request(q, from, to, BaiduTranslateApi.BAIDU_TRANSLATE_APPID, salt, sign)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(subscriber)
    }

    companion object {
        private var instance: Retrofit? = null

        @Synchronized
        fun getInstance(): Retrofit {
            if (instance == null) {
                instance = Retrofit.Builder()
                    .baseUrl(BaiduTranslateApi.BAIDU_TRANSLATE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                    .build()
            }
            return instance!!
        }
    }

}
